### ISSUE TITLE
*-mingw-dynamic triplet automatic detection does not work

### ISSUE THREAD

**Issue by OLEGSHA** [microsoft/vcpkg#30776]
**Describe the bug**
vcpkg toolchain for CMake fails to detect MinGW in two common use cases: when using Visual Studio 2022 configuration preset "Mingw64-Debug" and when using command line. `*-windows` is selected instead.

**Environment**
- OS: Windows 11
- vcpkg: 2023-03-29-664f8bb619b752430368d0f30a8289b761f5caba, PATH unmodified, `vcpkg integrate install` executed.
- MinGW: [w64devkit](https://github.com/skeeto/w64devkit) 1.18.0, GCC 11.2.0, installed into C:\msys64\mingw64\, PATH unmodified.
- CMake: 3.25.1-msvc1 when using VS2022 and 3.26.3, PATH unmodified, when using command line.

**To Reproduce**
A dummy CMake project is enough to trigger this problem:
```cmake
cmake_minimum_required(VERSION 3.12)
project(test)
```
Steps to reproduce the behavior with Visual Studio:
1. Open project.
2. In Manage Configurations, add "Mingw64-Debug".
3. Generate CMake files.
4. Check `CMakeCache.txt` to see `VCPKG_TARGET_TRIPLET` is `x64-windows`.

Steps to reproduce the behavior from command line:
```cmd
set Path=C:\msys64\mingw64\bin\;%Path%
cmake -S . -B build -G "MinGW Makefiles" -DCMAKE_TOOLCHAIN_FILE=<vcpkg root>/scripts/buildsystems/vcpkg.cmake
```
Check `build\CMakeCache.txt` to see `VCPKG_TARGET_TRIPLET` is `x64-windows`.

**Expected behavior**
`VCPKG_TARGET_TRIPLET` should be set to `x64-mingw-dynamic`.

I assume this is intended behavior from [vcpkg.cmake:342](https://github.com/microsoft/vcpkg/blob/45b0e4b5854bae6de9bfa1eeffbbad51bc1d19fe/scripts/buildsystems/vcpkg.cmake#L342).

**Failure logs**
No relevant log files. No error messages appear. CMake logs show that the compiler is indeed gcc in both scenarios.

**Additional context**
vcpkg does indeed act on this incorrect triplet. As to be expected from `VCPKG_TARGET_TRIPLET = x64-windows`, packages installed for MinGW are not available.

`vcpkg.cmake` attempts to detect MinGW by checking CMake variable `MINGW`. However, as adding a simple
```cmake
if (MINGW)
    message(WARNING "==== MinGW detected! ====")
endif()
```
demonstrates, when inserted right before the triplet detection code, that `MINGW` is not set when the toolchain file is executed. Inserting this check anywhere in `CMakeLists.txt`, however, shows that `MINGW` becomes set when user code begins execution. It seems logical to me that as toolchains may affect the choice of the compiler, CMake doesn't run MinGW detection until after toolchain file is completed.

It was suggested to me by @autoantwort that adding a check for `MinGW Makefiles` generator in `vcpkg.cmake` might solve this issue. However, this will fail to trigger if a different MinGW-compatible generator is used. In particular, this does not cover the Visual Studio case since it uses Ninja as the default generator in that preset.

**Comment by OLEGSHA** [comment]
Sorry for confusion, I will recreate this issue in the [proper repo](https://github.com/microsoft/vcpkg-tool).

**Comment by autoantwort** [comment]
> In particular, this does not cover the Visual Studio case since it uses Ninja as the default generator in that preset. 

Do you have a suggestion how mingw could be detected in this situation?

**Comment by OLEGSHA** [comment]
I'm not too well-versed in how CMake internals work. I don't have any immediately implementable solution. There are a several hacks I could come up with, like checking compiler name for one, but they all have obvious flaws, which would lead to weird inconsistencies that could seriously confuse users.

In my opinion the only robust solution is to rely on CMake to detect that for us.

## Option 1: delay triplet detection
Triplet detection could theoretically be deferred until CMake comes up with an answer naturally. For example, it could be triggered inside the various `find_*` modules - they will definitely execute after CMake has initialized. However, I'm pretty sure some configuration vcpkg performs is purely passive and must work immediately after the toolchain file is processed. In addition, this would result in a massive overhaul of the CMake integration structure, which does not look realistic.

## Option 2: trigger CMake detection early
CMake's internal mechanism for detecting compilers should be available at runtime, and it may be possible to run it manually from vcpkg. This is easier to implement, but it may carry unintended side effects.

**Comment by JonLiu1993** [comment]
This issue hasn’t been updated in 3 month, if it is still an issue, please reopen this issue.

**Comment by HunterZ** [comment]
I tried to report something similar via #23607. The problem is this line:
https://github.com/microsoft/vcpkg/blob/45b0e4b5854bae6de9bfa1eeffbbad51bc1d19fe/scripts/buildsystems/vcpkg.cmake#L350

Specifically, it doesn't work because CMake itself doesn't populate that variable until sometime after the toolchain file processing stage.

Unfortunately my report was ultimately ignored, so I had to settle for a workaround of explicitly specifying the triplet in my development environment.
